pintc/src/expr.rs

```diff
@@ -1103,3 +895,476 @@ impl Expr {
+) -> bool {
+    lhs_expr.get(contract).eq(contract, rhs_expr.get(contract))
+        && lhs_index
+            .get(contract)
+            .eq(contract, rhs_index.get(contract))
+}
+
+pub fn tuple_field_access_eq(
+    contract: &Contract,
+    lhs_tuple: &ExprKey,
+    lhs_field: &TupleAccess,
+    rhs_tuple: &ExprKey,
+    rhs_field: &TupleAccess,
+) -> bool {
+    lhs_tuple
+        .get(contract)
+        .eq(contract, rhs_tuple.get(contract))
+        && match (lhs_field, rhs_field) {
+            (TupleAccess::Error, TupleAccess::Error) => true,
+
+            (TupleAccess::Index(lhs_index), TupleAccess::Index(rhs_index)) => {
+                lhs_index == rhs_index
+            }
+
+            (TupleAccess::Name(lhs_name), TupleAccess::Name(rhs_name)) => lhs_name == rhs_name,
+
+            _ => false,
+        }
+}
+
+pub fn cast_eq(
+    contract: &Contract,
+    lhs_value: &ExprKey,
+    lhs_ty: &Type,
+    rhs_value: &ExprKey,
+    rhs_ty: &Type,
+) -> bool {
+    lhs_value
+        .get(contract)
+        .eq(contract, rhs_value.get(contract))
+        && lhs_ty.eq(contract, rhs_ty)
+}
+
+pub fn in_eq(
+    contract: &Contract,
+    lhs_value: &ExprKey,
+    lhs_collection: &ExprKey,
+    rhs_value: &ExprKey,
+    rhs_collection: &ExprKey,
+) -> bool {
+    lhs_value
+        .get(contract)
+        .eq(contract, rhs_value.get(contract))
+        && lhs_collection
+            .get(contract)
+            .eq(contract, rhs_collection.get(contract))
+}
+
+pub fn range_eq(
+    contract: &Contract,
+    lhs_lb: &ExprKey,
+    lhs_ub: &ExprKey,
+    rhs_lb: &ExprKey,
+    rhs_ub: &ExprKey,
+) -> bool {
+    lhs_lb.get(contract).eq(contract, rhs_lb.get(contract))
+        && lhs_ub.get(contract).eq(contract, rhs_ub.get(contract))
+}
+
+pub fn generator_eq(
+    contract: &Contract,
+    lhs_kind: &GeneratorKind,
+    lhs_gen_ranges: &Vec<(Ident, ExprKey)>,
+    lhs_conditions: &Vec<ExprKey>,
+    lhs_body: &ExprKey,
+    rhs_kind: &GeneratorKind,
+    rhs_gen_ranges: &Vec<(Ident, ExprKey)>,
+    rhs_conditions: &Vec<ExprKey>,
+    rhs_body: &ExprKey,
+) -> bool {
+    lhs_kind == rhs_kind
+        && rhs_gen_ranges.len() == lhs_gen_ranges.len()
+        && lhs_gen_ranges
+            .iter()
+            .enumerate()
+            .all(|(i, (lhs_ident, lhs_gen_range))| {
+                let (rhs_ident, rhs_gen_range) = &rhs_gen_ranges[i];
+
+                lhs_ident == rhs_ident
+                    && lhs_gen_range
+                        .get(contract)
+                        .eq(contract, rhs_gen_range.get(contract))
+            })
+        && lhs_conditions.len() != rhs_conditions.len()
+        && lhs_conditions.iter().enumerate().all(|(i, lhs_condition)| {
+            lhs_condition
+                .get(contract)
+                .eq(contract, rhs_conditions[i].get(contract))
+        })
+        && lhs_body == rhs_body
+}
+
+pub fn map_eq(
+    contract: &Contract,
+    lhs_param: &Ident,
+    lhs_range: &ExprKey,
```

Comment:
I could be wrong but it _looks_ like slotmap keys are `Copy` by default: https://docs.rs/slotmap/latest/src/slotmap/lib.rs.html#456 (and https://docs.rs/slotmap/latest/src/slotmap/lib.rs.html#260)

perhaps no `&` is needed here?

Reply:
Yep, they're definitely (and very handily) `Copy.`